cmd/kubeadm/app/cmd/config.go

```diff
@@ -270,15 +272,62 @@ func NewCmdConfigMigrate(out io.Writer) *cobra.Command {
 func NewCmdConfigUpload(out io.Writer, kubeConfigFile *string) *cobra.Command {
 	cmd := &cobra.Command{
 		Use:   "upload",
-		Short: "Upload configuration about the current state, so that 'kubeadm upgrade' can later know how to configure the upgraded cluster.",
+		Short: "Uploads the kubeadm and kubelet configuration to a ConfigMap. This is a simple shorthand for the upload-config init phase.",
+		Long:  cmdutil.MacroCommandLongDescription,
 		RunE:  cmdutil.SubCmdRunE("upload"),
 	}
 
-	cmd.AddCommand(NewCmdConfigUploadFromFile(out, kubeConfigFile))
+	// Add legacy deprecated sub-command
 	cmd.AddCommand(NewCmdConfigUploadFromFlags(out, kubeConfigFile))
+
+	// We actually need only the init cmd here, the fake kubeadm cmd is used only for generating correct help messages
+	fakeKubeadmCmd := &cobra.Command{Use: "kubeadm"}
+	fakeKubeadmCmd.AddCommand(NewCmdInit(out, nil))
+
+	// Wrap the upload config init sub-phases into sub-commands
+	uploadConfigPhase := initphases.NewUploadConfigPhase()
+	for _, phase := range uploadConfigPhase.Phases {
+		cmd.AddCommand(configUploadPhaseToCommand(fakeKubeadmCmd, &phase, kubeConfigFile))
+	}
+
+	return cmd
+}
+
+// configUploadPhaseToCommand creates a wrapper cobra command for a given upload-config sub-phase
+func configUploadPhaseToCommand(baseCmd *cobra.Command, phase *phaseworkflow.Phase, kubeConfigFile *string) *cobra.Command {
+	cmd := &cobra.Command{
+		Use:                phase.Name,
+		Short:              phase.Short,
+		Long:               phase.Long,
+		Example:            phase.Example,
+		Aliases:            phase.Aliases,
+		DisableFlagParsing: true,
+		Run: func(cmd *cobra.Command, args []string) {
+			runConfigUploadPhase(baseCmd, cmd.Name(), args, *kubeConfigFile)
+		},
+	}
+
+	// setup the extra "from-file" alias to the "kubeadm" phase command
+	if phase.Name == "kubeadm" {
+		cmd.Aliases = append([]string{}, cmd.Aliases...)
+		cmd.Aliases = append(cmd.Aliases, "from-file")
+	}
+
 	return cmd
 }
 
+func runConfigUploadPhase(baseCmd *cobra.Command, phaseName string, args []string, kubeConfigFile string) {
+	phaseArgs := []string{"init", "phase", "upload-config", phaseName}
+	// If we are given a specific kubeconfig path by the parent commands, prepend that.
+	// That way it will be taken into account or overwritten if the user specified another one in args.
+	if kubeConfigFile != "" {
+		phaseArgs = append(phaseArgs, "--"+options.KubeconfigPath, kubeConfigFile)
+	}
+	phaseArgs = append(phaseArgs, args...)
+	baseCmd.SetArgs(phaseArgs)
+	baseCmd.Execute()
```

Comment:
This seems really weird.  At the end of the day there should be a common entry point vs. plumbing back through cobra.